src/cargo/util/toml/mod.rs

```diff
@@ -1944,8 +1964,6 @@ fn detailed_dep_to_dependency<P: ResolveToPath + Clone>(
     }
 
     if let Some(p) = orig.public {
-        cx.features.require(Feature::public_dependency())?;
-
         if dep.kind() != DepKind::Normal {
             bail!("'public' specifier can only be used on regular dependencies, not {:?} dependencies", dep.kind());
```

Comment:
Is this before or after we set `public = false`?

If before, we should probably check the unstable status and report this as a warning that this will fail when stabilized.  Or in other words, without the `-Z`, users should get the experience they have before any work was done on this but be warned what will break

Reply:
I had migrated the  code to here, because `detailed_dep_to_dependency` will invoke by `read_manifest_from_str`.

Reply:
> If before, we should probably check the unstable status and report this as a warning that this will fail when stabilized. 

We still aren't downgrading this `bail` to a warning for stable.

Reply:
> CARGO_RUN_BUILD_STD_TESTS=true cargo test -p cargo --test build-std

the `build-std` will fail, because they don't pass `-Zpublic-dependency`, detail :https://github.com/rust-lang/cargo/actions/runs/7665093719/job/20890409198?pr=13340

Reply:
Had addressed.

Reply:
I'm not seeing how.  Could you clarify?  Depending on order, we either fire off the error or we are silent.  We should probably have a test for this.

Reply:
For stable, `bail_public_on_stable` can verify only emit error on stable.
For nightly,  `verify_z_public_dependnecy` can verify emit  warn if  no `-Zpublic-dependency` and keep silenc if pass `-Zpublic-dependency` on nightly.

I think the only thing that's different now is that the nightly version displays a warning instead of a direct error.

Reply:
Can you add a variant of the `pub_dev_dependency` test that runs without the unstable feature being enabled?

Reply:
Thanks for the heads-up. Similar tests have been added.

(`pub_dev_dependency_without_feature_in_stable` and `pub_dev_dependency_without_feature_in_nightly`)

Reply:
So those tests show that we are erroring on this unstable feature when we should instead warn that this will error out in the future when its stabilized.

I will say its a separate question of whether its worth fixing (depends on how intrusive the fix is).  The general philisophy of this approach to unstable features is they should act like you specified the keys back in <1.30.0 but warn people if there will be a breakage on stabilization.

Reply:
em, it seems to be blocking here.
If you look at the logic here again, both the stable and nightly versions are throwing errors.
Can we keep throwing errors on nightly if no -Z or features enabled?

Reply:
Generally, code activates with the feature being enabled, not just if nightly is used.

Reply:
Please see https://github.com/rust-lang/cargo/pull/13340/files#diff-3d8669050f6f1c07884ef5a4e9d344a98c7270f73f4683125a00e8a3cc6381e6R1963.

Reply:
Now that we have a test, I'm moving the discussion to https://github.com/rust-lang/cargo/pull/13340/files#r1498114047